internal/resource/state.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package resource
+
+import (
+	"github.com/juju/errors"
+)
+
+// These are the valid resource states.
+const (
+	stateUnknown State = iota
+
+	// StateAvailable represents a resource which will be used by any units at
+	// this point in time
+	StateAvailable
+
+	// StatePotential indicates there is a different revision of the resource available
+	// in a repository. Used to let users know a resource can be upgraded.
+	StatePotential
+)
+
+var states = map[State]string{
+	StateAvailable: "available",
+	StatePotential: "potential",
+}
+
+// State identifies the resource state in an application
+type State int
+
+// ParseState converts the provided string into an State.
+// If it is not a known state then an error is returned.
+func ParseState(value string) (State, error) {
+	for o, str := range states {
+		if value == str {
+			return o, nil
+		}
+	}
+	return stateUnknown, errors.Errorf("unknown state %q", value)
+}
+
+// String returns the printable representation of the state.
+func (o State) String() string {
+	return states[o]
+}
+
+// Validate ensures that the state is correct.
+func (o State) Validate() error {
+	// Ideally, only the (unavoidable) zero value would be invalid.
+	// However, typedef'ing int means that the use of int literals
+	// could result in invalid Type values other than the zero value.
```

Comment:
One reason I prefer to use typed strings and no iota for things like this. It also make debugging a lot easier when looking at raw data.

We do not consensus within the team on this one.

Reply:
Done